Separate the dust threshold from the minrelaytxfee

ptschip opened the pull request:
ABC has changed their default minRelayTxFee to 250 sat/byte and while changing it in BU doesn't really change much because the mempool limiter will adjust this value on the fly anyway, but I thought it would be a good idea to align ourselves with ABC.  Then we should probably review how we want the mempool limiting defaults to work in the future so we don't end up trampling on zero-conf.

So do make the change we first need to separate out the dustrelayfee from the minrelaytxfee or we end up having issues with the tests but also it's a good idea that users can change the dustrelayfee on the fly if needed in the future.  This was mostly a port from Core with some minor changes so we don't shadow any variables.  After this one done it was a simple one liner to change the minrelaytxfee default.

morcos committed:
Introduce -dustrelayfee

ptschip committed:
Do no shadow dustRelayFee in IsDust() and GetDustThreshHold()

Also remove references to minRelayTxFee as they are no longer
relevant.

ptschip committed:
Fee tolerance should be at least one satoshi

This accounts for any rounding errors when minrelaytxfee is less
than 500 satoshis.

sickpig reviewed (commented):
@ptschip I've just looked at ABC code and I see no change in regards to the default value of minimum relay fee per KB. They still have it set at 1000 sat per bytes or so it seems to me: 

```
src/validation.h:static const Amount DEFAULT_MIN_RELAY_TX_FEE(1000);
```

ptschip commented:
here's where they change their tx relay fee... https://reviews.bitcoinabc.org/rABCa3d4a2a83e479cebfc287d39fa5d2385939faff8

ptschip committed:
Remove dustRelayFee and the calculation for dust

Rely on a simple dust threshold value which can be configured
by the command line if needed.

Update comments.

Also update the command line help text to be more clear

Turn nDustThreshold into a Tweak

ptschip commented:
got rid of the dustrelayfee and rely simply on a default_dust_threshold.  Also minrelaytxfee was reverted back to 1000sat/Kb

sickpig commented on src/policy/policy.h:
```diff
@@ -30,6 +30,13 @@ static const unsigned int MAX_STANDARD_TX_SIGOPS = BLOCKSTREAM_CORE_MAX_BLOCK_SI
+ * minRelayTxFee, however changing the dust limit changes which transactions are
+ * standard and should be done with care and ideally rarely. It makes sense to
+ * only increase the dust limit after prior releases were already not creating
+ * outputs below the new threshold */
```
I'd also add the reasoning behind choosing 546, i.e. the comment remove from `GetDustThreshold()`:

```c++
        // "Dust" is defined in terms of CTransaction::minRelayTxFee,
        // which has units satoshis-per-kilobyte.
        // If you'd pay more than 1/3 in fees
        // to spend something, then we consider it dust.
        // A typical spendable txout is 34 bytes big, and will
        // need a CTxIn of at least 148 bytes to spend:
        // so dust is a spendable txout less than
        // 546*minRelayTxFee/1000 (in satoshis)
```

sickpig commented on src/primitives/transaction.h:
```diff
@@ -144,24 +146,12 @@ class CTxOut
+        return (CAmount)nDustThreshold;
     }
 
-    bool IsDust(const CFeeRate &minRelayTxFee) const { return (nValue < GetDustThreshold(minRelayTxFee)); }
```
didn't we just change the semantics of this functions? I mean before the threshold was dependent on tx ser size now is always 546 satoshi.

ptschip replied:
No not really, it always was 546 satoshi.  After all those calculations and getting of the serialize size, the result was the same  1000/1000 * 546.

sickpig replied:
need to look at the code more closely. but the code before returned `false` in case `scriptPubKey.IsUnspendable()` returned `true`.

This is the code of method: 

```c++
    bool IsUnspendable() const
    {
         return (size() > 0 && *begin() == OP_RETURN) || (size() > MAX_SCRIPT_SIZE);
    }
```
Do we check for this condition with the current proposal?

sickpig commented:
@ptschip `fundrawtransaction.py` fail regularly on travis

```
fundrawtransaction.py          | False  | 29 s
```

gandrewstone commented:
also run formatting

gandrewstone commented on src/allowed_args.cpp:
```diff
@@ -561,6 +561,8 @@ static void addNodeRelayOptions(AllowedArgs &allowedArgs)
             strprintf(_("Maximum size of data in data carrier transactions we relay and mine (default: %u)"),
                     MAX_OP_RETURN_RELAY))
+        .addArg("dust-threshold=<amt>", requiredAmount,
+            strprintf(_("Dust Threshold (in satoshis) used to defined dust. (default: %s)"), DEFAULT_DUST_THRESHOLD))
```
please don't use the word in the definition.  How about:

"Dust Threshold (in satoshis) defines the minimum quantity an output may contain for the transaction to be considered 'standard' (relayable)"

ptschip replied:
ok, my bad...will change that.

gandrewstone commented on src/init.cpp:
```diff
@@ -798,6 +798,10 @@ bool AppInit2(Config &config, boost::thread_group &threadGroup, CScheduler &sche
             strprintf("acceptnonstdtxn is not currently supported for %s chain", chainparams.NetworkIDString()));
+
+    // Set Dust Threshold for outputs.
+    nDustThreshold = GetArg("-dust-threshold", DEFAULT_DUST_THRESHOLD);
```
This would be better as a tweak since it should be dynamically adjustable.  Also, no other parameter uses word-word.  like "bytespersigop" we should call this dustthreshold.  I'm concerned that the - may screw up other command line/conf file parsers, even if it works for ours.

ptschip replied:
ok will do

ptschip commented:
yes i think you're correct...i'll have to put that in again, good catch.

On 27/07/2018 6:27 AM, Andrea Suisani wrote:
>
> *@sickpig* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/primitives/transaction.h
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1181#discussion_r205771729>:
>
> >      }
>  
> -    bool IsDust(const CFeeRate &minRelayTxFee) const { return (nValue < GetDustThreshold(minRelayTxFee)); }
>
> need to look at the code more closely. but the code before returned
> |false| in case |scriptPubKey.IsUnspendable()| returned |true|.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1181#discussion_r205771729>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0CFR6DjP_Hm2tyKIVB4pnhYGWNmKks5uKxUtgaJpZM4VCuhO>.
>
